src/motor_hardware.cc

```diff
@@ -664,12 +764,92 @@ void MotorDiagnostics::motor_power_status(DiagnosticStatusWrapper &stat) {
+    }
+    stat.summary(DiagnosticStatusWrapper::OK, option_descriptions);
 }
 
+// i2c_BufferRead()   A host OS system specific  utility to open, read, close from an I2C device
+//
+// The I2C address is the 8-bit address which is the 7-bit addr shifted left in some code
+// If chipRegAddr is greater than 1 we write this out for the internal chip address for the following read(s)
+//
+// Returns number of bytes read where 0 or less implies some form of failure
+//
+// NOTE: The i2c8bitAddr will be shifted right one bit to use as 7-bit I2C addr
+//
+static int i2c_BufferRead(const char *i2cDevFile, uint8_t i2c8bitAddr, 
+                          uint8_t *pBuffer, int16_t chipRegAddr, uint16_t NumByteToRead)
+{
+   int bytesRead = 0;
+   int retCode   = 0;
+
+   // If your system requires a lock on the bus, it should be called here 
+   // if ((semId >= 0) && (ipc_sem_lock(semId) < 0)) {
+   //   // printf("i2c_BufferRead: Cannot obtain I2C lock!  ERROR: %s\n", strerror(errno));
+   //   return -1;
+   // }
+
+    // we are now free to access the I2C hardware
+    int fd;                                         // File descrition
+    int  address   = i2c8bitAddr >> 1;              // Address of the I2C device
+    unsigned char buf[8];                           // Buffer for data being written to the i2c device
+
+    if ((fd = open(i2cDevFile, O_RDWR)) < 0) {        // Open port for reading and writing
+      retCode = -2;
+      ROS_ERROR("Cannot open I2C def of %s with error %s", i2cDevFile, strerror(errno));
+      goto exitWithSemUnlock;
+    }
+
+    // The ioctl here will address the I2C slave device making it ready for 1 or more other bytes
+    if (ioctl(fd, I2C_SLAVE, address) < 0) {        // Set the port options and addr of the dev
+      retCode = -3;
+      ROS_ERROR("Failed to get bus access to I2C device %s!  ERROR: %s", i2cDevFile, strerror(errno));
+      goto exitWithFileCloseAndSemUnlock;
+    }
+
+    if (chipRegAddr < 0) {     // Suppress reg address if negative value was used
+      buf[0] = (uint8_t)(chipRegAddr);          // Internal chip register address
+      if ((write(fd, buf, 1)) != 1) {           // Write both bytes to the i2c port
+        retCode = -4;
+        goto exitWithFileCloseAndSemUnlock;
+      }
+    }
+
+    // Now we have to read from the chip as the register start address was just setup
+    // if (ioctl(fd, I2C_SLAVE, address) < 0) {        // Set the port options and addr of the dev
+    //   ROS_INFO("%s: Failed to get bus access to I2c port!  ERROR: %s\n", THIS_NODE_NAME,  strerror(errno));
+    //   goto exitWithFileCloseAndSemUnlock;
+    // }
+
+    bytesRead = read(fd, pBuffer, NumByteToRead);
+    if (bytesRead != NumByteToRead) {      // verify the number of bytes we requested were read
+      retCode = -9;
+      goto exitWithFileCloseAndSemUnlock;
+    }
+    retCode = bytesRead;
+
+  exitWithFileCloseAndSemUnlock:
+    close(fd);
+
+  exitWithSemUnlock:
```

Comment:
Another question I had for you.  As I said, this comes from my own code where I used this.  I left inklings of it here so I could ask you about it.    My understanding is in ROS different nodes cannot be running at the same time so a lock for I2C is sort of automatic in ROS.   If you feel that is a true statement I'll take out all the lock stuff, even the bits in commented out source.     Again, I meant to ask about this so 'there you are'.

Reply:
The whole point of ROS is that you do have multiple nodes operating in parallel. Within the same node, the threading model is up to you. I will have to think about the i2c stuff a bit more, but we can do without a lock for now (since nobody else is currently using i2c along side us).

Reply:
Looking into this some more. the kernel provides a cleaner atomic interface for a I2C register read operation using the I2C_RDWR ioctl. This does all the reading and writing in one step atomically, so that no other processes or kernel modules can interrupt it.

We should use this functionality. As it stands this code is brittle to other systems trying to use I2C at the same time (including the kernel RTC module).

See I2C_RDWR in here https://www.kernel.org/doc/Documentation/i2c/dev-interface
And some example code here: https://gist.github.com/JamesDunne/9b7fbedb74c22ccc833059623f47beb7